src/compas/geometry/curves/nurbs.py

```diff
@@ -0,0 +1,462 @@
+from math import sqrt
+
+from compas.geometry import Point
+from compas.geometry import Frame
+from compas.utilities import linspace
+
+from ._curve import Curve
+
+
+class NurbsCurve(Curve):
+    """Class representing a NURBS curve.
+
+    Attributes
+    ----------
+    points: List[Point]
+        The control points of the curve.
+    weights: List[float]
+        The weights of the control points.
+    knots: List[float]
+        The knot vector, without duplicates.
+    multiplicities: List[int]
+        The multiplicities of the knots in the knot vector.
+    knotsequence: List[float]
+        The knot vector, with repeating values according to the multiplicities.
+    degree: int
+        The degree of the polynomials.
+    order: int
+        The order of the curve.
+    domain: Tuple[float, float]
+        The parameter domain.
+    start: :class:`Point`
+        The point corresponding to the start of the parameter domain.
+    end: :class:`Point`
+        The point corresponding to the end of the parameter domain.
+    is_closed: bool
+        True if the curve is closed.
+    is_periodic: bool
+        True if the curve is periodic.
+    is_rational: bool
+        True is the curve is rational.
+    length: float
+        Length of the curve.
+
+    """
+
+    @property
+    def DATASCHEMA(self):
+        from schema import Schema
+        from compas.data import is_float3
+        from compas.data import is_sequence_of_int
+        from compas.data import is_sequence_of_float
+        return Schema({
+            'points': lambda points: all(is_float3(point) for point in points),
+            'weights': is_sequence_of_float,
+            'knots': is_sequence_of_float,
+            'multiplicities': is_sequence_of_int,
+            'degree': int,
+            'is_periodic': bool
+        })
+
+    @property
+    def JSONSCHEMANAME(self):
+        raise NotImplementedError
+
+    def __init__(self, name=None):
+        super().__init__(name=name)
```

Comment:
Does this not need to be ipy friendly? Or is the intention that compas nurbs curves and surfaces can only pass in and out of rhino in the form of `data`?

Reply:
i was thinking of the latter, but we could also go for a different approach using pluggable `__new__` with plugins from Rhino and OCC as mentioned above...